### PR TITLE
TEST: Replacing hardcoded mmap with dynamic

### LINKED ISSUE #8827: test_ucp_mmap.fixed failure with CUDA 12 + ROCE
### Describe the bug
UCX gtest test_ucp_mmap.fixed fails on platforms with A40 GPU and CUDA12. I added additional log messages to the test.
I believe the problem is that this specific test is using *hardcoded*  `0xff0000000` vaddr and passing it as MAP_FIXED. Somehow it collides with `cudaHostRegister`.  If I run just single test I cannot reproduce the problem. I had to run full gtest to get it reproduced. 

```
[       OK ] dcx/test_ucp_mmap.reg_advise/2 (232 ms)
[ RUN      ] dcx/test_ucp_mmap.reg_advise/3 <dc_x,cuda_copy,rocm_copy/no_rcache>
gtest debug >>> cudaHostRegister 0x7fff08239670, 1, 1
gtest debug >>> cudaHostRegister 0x7fff0823b66f, 1, 1
gtest debug >>> cudaHostRegister 0x7f977effd010, 134217728, 1
[       OK ] dcx/test_ucp_mmap.reg_advise/3 (218 ms)
[ RUN      ] dcx/test_ucp_mmap.fixed/0 <dc_x,cuda_copy,rocm_copy>
gtest debug >>> cudaHostRegister 0x7fff08239670, 1, 1
gtest debug >>> cudaHostRegister 0x7fff0823b66f, 1, 1
gtest debug >>> cudaHostRegister 0xff0000000, 4096, 1
[1673977958.225280] [WS-RTX:173118:0]    cuda_copy_md.c:156  UCX  ERROR cudaHostRegister(address, length, 0x01)() failed: invalid argument
[1673977958.225287] [WS-RTX:173118:0]          ucp_mm.c:62   UCX  ERROR failed to register address 0xff0000000 (host) length 4096 on md[2]=cuda_cpy: Input/output error (md supports: host|cuda|cuda-managed)
../../../test/gtest/ucp/test_ucp_mmap.cc:866: Failure
Error: Input/output error
../../../test/gtest/common/test.cc:366: Failure
Failed
Got 2 errors and 0 warnings during the test
```

### Steps to Reproduce
- Command line: `make -C test/gtest test
- UCX version used: master
- **Any UCX environment variables used** - none

### Setup and versions
- OS version ubuntu 22.04, inbox rdma-core, CUDA 12 (A6000), x86_64

### PR DESCRIPTION
**Pull request by shamisp**
Fixes #8827

### CODE REVIEW

File: test/gtest/common/test_helpers.cc [hunk 1]
```diff
@@ -606,8 +606,20 @@ uint16_t get_port() {
     return port;
 }
 
-void *mmap_fixed_address() {
-    return (void*)0xff0000000;
+void *mmap_fixed_address(size_t length) {
+    void *ret = NULL;
+    void *ptr = mmap(NULL, length, PROT_READ|PROT_WRITE,
+                     MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
+
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:613]: remove spaceline

File: test/gtest/common/test_helpers.cc [hunk 2]
```diff
@@ -606,8 +606,20 @@ uint16_t get_port() {
...
+    if (ptr == MAP_FAILED) {
+        goto out;
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:615]: IMO would be simpler to just return NULL here, and return ptr at line 620

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 3]
```diff
@@ -850,7 +850,11 @@ UCS_TEST_P(test_ucp_mmap, fixed) {
 
     for (int i = 0; i < 1000 / ucs::test_time_multiplier(); ++i) {
         size_t size = (i + 1) * ((i % 2) ? 1000 : 1);
-        void *ptr = ucs::mmap_fixed_address();
+        void *ptr = ucs::mmap_fixed_address(size);
+
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc:854]: remove space line

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 4]
```diff
@@ -850,7 +850,11 @@ UCS_TEST_P(test_ucp_mmap, fixed) {
...
+        if (ptr == NULL) {
+            UCS_TEST_SKIP_R("mmap failed to allocate memory region");
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc:856]: IMO we should fail the test

File: test/gtest/uct/test_mem.cc [hunk 5]
```diff
@@ -219,9 +223,13 @@ UCS_TEST_P(test_mem, mmap_fixed) {
     const size_t            page_size   = ucs_get_page_size();
     const size_t            n_tryes     = 101;
     uct_alloc_method_t      meth;
-    void*                   p_addr      = ucs::mmap_fixed_address();
     size_t                  length      = 1;
     size_t                  n_success;
+    void*                   p_addr      = ucs::mmap_fixed_address(page_size * 2 * n_tryes);
+
+    if (p_addr == NULL) {
+        UCS_TEST_SKIP_R("mmap failed to allocate memory region");
```
Comment (yosefe) [inline test/gtest/uct/test_mem.cc:231]: IMO we should fail the test

**Comment by shamisp** [comment]
@yosefe -I believe I addressed the above concerns

File: test/gtest/common/test_helpers.cc [hunk 6]
```diff
@@ -606,8 +606,18 @@ uint16_t get_port() {
...
 }
 
-void *mmap_fixed_address() {
-    return (void*)0xff0000000;
+void *mmap_fixed_address(size_t length) {
+    void *ret = NULL;
+    void *ptr = mmap(NULL, length, PROT_READ | PROT_WRITE, 
+                     MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (ptr == MAP_FAILED) {
+        return ret;
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:614]: return NULL

File: test/gtest/common/test_helpers.cc [hunk 7]
```diff
@@ -606,8 +606,18 @@ uint16_t get_port() {
...
+    }
+
+    ret = ptr;
+    munmap(ptr, length);
+
+    return ret;
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc:620]: IMO ret variable is redundant - can just return ptr here

File: test/gtest/uct/test_mem.cc (hunk 5 above)
Comment (yosefe) [reply test/gtest/uct/test_mem.cc:230]: unresolved

**Comment by shamisp** [comment]
@yosefe fixed

**Review by yosefe — COMMENTED** [review]
LGTM, but the Coverity test failure needs to be fixed by a suppression `/* coverity[use_after_free] */` before "return ptr"
```
+ cov-format-errors --dir /home/swx-azure-svc_azpcontainer/57621-20230130.13/build/cov_build_devel --emacs-style
/__w/1/s/test/gtest/common/test_helpers.cc:618
  Type: Use after free (USE_AFTER_FREE)

/__w/1/s/test/gtest/common/test_helpers.cc:612:
  1. path: Condition "ptr == (void *)0xffffffffffffffff", taking false branch.
/__w/1/s/test/gtest/common/test_helpers.cc:616:
  2. freed_arg: "munmap" frees "ptr".
/__w/1/s/test/gtest/common/test_helpers.cc:618:
  3. use_after_free: Using freed pointer "ptr".

```

File: test/gtest/common/test_helpers.cc [hunk 8]
```diff
@@ -606,8 +606,16 @@ uint16_t get_port() {
     return port;
 }
 
-void *mmap_fixed_address() {
-    return (void*)0xff0000000;
+void *mmap_fixed_address(size_t length) {
+    void *ptr = mmap(NULL, length, PROT_READ | PROT_WRITE,
+                     MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    if (ptr == MAP_FAILED) {
+        return NULL;
```
Comment (rakhmets) [inline test/gtest/common/test_helpers.cc:613]: You can use `nullptr` in C++ code.
```suggestion
        return nullptr;
```

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 9]
```diff
@@ -850,7 +850,10 @@ UCS_TEST_P(test_ucp_mmap, fixed) {
 
     for (int i = 0; i < 1000 / ucs::test_time_multiplier(); ++i) {
         size_t size = (i + 1) * ((i % 2) ? 1000 : 1);
-        void *ptr = ucs::mmap_fixed_address();
+        void *ptr   = ucs::mmap_fixed_address(size);
+        if (ptr == NULL) {
```
Comment (rakhmets) [inline test/gtest/ucp/test_ucp_mmap.cc:854]:
```suggestion
        if (ptr == nullptr) {
```

File: test/gtest/uct/test_mem.cc [hunk 10]
```diff
@@ -139,17 +139,20 @@ UCS_TEST_P(test_mem, md_alloc) {
...
+    uct_md_h                 md;
+    unsigned                 j;
+
+    const size_t             page_size   = ucs_get_page_size();
+    const size_t             n_tryes     = 101;
+    uct_alloc_method_t       meth;
+    size_t                   length      = 1;
+    size_t                   n_success;
+    void*                    p_addr      = ucs::mmap_fixed_address(page_size * 2 * n_tryes);
+    if (p_addr == NULL) {
```
Comment (rakhmets) [inline test/gtest/uct/test_mem.cc:153]:
```suggestion
    if (p_addr == nullptr) {
```

File: test/gtest/uct/test_mem.cc [hunk 11]
```diff
@@ -219,9 +223,12 @@ UCS_TEST_P(test_mem, mmap_fixed) {
     const size_t            page_size   = ucs_get_page_size();
     const size_t            n_tryes     = 101;
     uct_alloc_method_t      meth;
-    void*                   p_addr      = ucs::mmap_fixed_address();
     size_t                  length      = 1;
     size_t                  n_success;
+    void*                   p_addr      = ucs::mmap_fixed_address(page_size * 2 * n_tryes);
+    if (p_addr == NULL) {
```
Comment (rakhmets) [inline test/gtest/uct/test_mem.cc:229]:
```suggestion
    if (p_addr == nullptr) {
```

**Review by rakhmets — DISMISSED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
failure seems relevant:
```
[ RUN      ] ud/test_ucp_mmap.fixed/0 <ud_v,cuda_copy,rocm_copy>
[swx-rdmz-ucx-new-02:22546:0:22546] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xafe7684)

/scrap/azure/agent-02/AZP_WORKSPACE/2/s/contrib/../src/ucs/datastruct/mpool.c: [ ucs_mpool_cleanup() ]
      ...
      161         elem = next_elem;
      162         VALGRIND_MAKE_MEM_DEFINED(elem, sizeof *elem);
      163         next_elem = elem->next;
==>   164         if (data->ops->obj_cleanup != NULL) {
      165             obj = elem + 1;
      166             VALGRIND_MEMPOOL_ALLOC(mp, obj, mp->data->elem_size - sizeof(ucs_mpool_elem_t));
      167             VALGRIND_MAKE_MEM_DEFINED(obj, mp->data->elem_size - sizeof(ucs_mpool_elem_t));

==== backtrace (tid:  22546) ====
 0 0x000000000005a410 ucs_mpool_cleanup()  /scrap/azure/agent-02/AZP_WORKSPACE/2/s/contrib/../src/ucs/datastruct/mpool.c:164
 1 0x0000000000090526 uct_ud_iface_t_cleanup()  /scrap/azure/agent-02/AZP_WORKSPACE/2/s/contrib/../src/uct/ib/ud/base/ud_iface.c:607
 2 0x0000000000090526 uct_ud_iface_t_cleanup()  /scrap/azure/agent-02/AZP_WORKSPACE/2/s/contrib/../src/uct/ib/ud/base/ud_iface.c:608
 3 0x000000000007e5ce ucs_class_call_cleanup_chain()  /scrap/azure/agent-02/AZP_WORKSPACE/2/s/contrib/../src/ucs/type/class.c:56
 4 0x0000000000097b70 uct_ud_verbs_iface_t_delete()  /scrap/azure/agent-02/AZP_WORKSPACE/2/s/contrib/../src/uct/ib/ud/verbs/ud_verbs.c:795
 5 0x0000000000067784 ucp_worker_uct_iface_close()  /scrap/azure/agent-02/AZP_WORKSPACE/2/s/contrib/../src/ucp/core/ucp_worker.c:801
```

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
```
023-02-02T07:01:20.8205441Z [ RUN      ] alloc_methods/test_mem.mmap_fixed/3 <4>
2023-02-02T07:01:20.8206534Z /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/uct/test_mem.cc:262: Failure
2023-02-02T07:01:20.8207642Z uct_mem_free failed to release memory region
2023-02-02T07:01:20.8208486Z [  FAILED  ] alloc_methods/test_mem.mmap_fixed/3, where GetParam() = 4 (0 ms)
```

File: test/gtest/uct/test_mem.cc [hunk 12]
```diff
@@ -247,10 +254,13 @@ UCS_TEST_P(test_mem, mmap_fixed) {
             EXPECT_EQ(meth, uct_mem.method);
             EXPECT_EQ(p_addr, uct_mem.address);
             EXPECT_GE(uct_mem.length, (size_t)1);
-            /* touch the page*/
+            /* touch the page */
             memset(uct_mem.address, 'c', uct_mem.length);
             EXPECT_EQ(*(char*)p_addr, 'c');
             status = uct_mem_free(&uct_mem);
+            if (status == UCS_OK) {
```
Comment (rakhmets) [inline test/gtest/uct/test_mem.cc:261]:
```suggestion
            if (status != UCS_OK) {
```

**Review by yosefe — DISMISSED** [review]

**Review by rakhmets — DISMISSED** [review]

**Review by rakhmets — APPROVED** [review]

**Pull request merged**